app_history.md

```diff
@@ -0,0 +1,423 @@
+
+## Goals
+
+Our primary goals are as follows:
+
+- Allow easy conversion of cross-document navigations into single-page app same-document navigations, without fragile hacks like a global `click` handler.
+
+- Provide a uniform way to signal single-page app navigations, including their duration.
+
+- Provide a reliable system to tie application and UI state to history entries.
+
+- Provide events for notifying the application about navigations through the list of history entries, which they can use to synchronize application or UI state.
+
+- Allow analytics (first- or third-party) to watch for navigations, including gathering timing information about how long they took, without interfering with the rest of the application.
+
+- Provide a way for an application to reliably navigate through its own history stack.
+
+- Provide a reasonable layering onto and integration with the existing `window.history` API, in terms of spec primitives and ensuring non-terrible behavior when both are used.
+
+Non-goals:
+
+- Provide applications knowledge of cross-origin history entries or state.
+
+- Provide applications knowledge of other frames' entries or state.
+
+- Handle the case where the Android back button is being used as a "modal close signal"; instead, we believe that's best handled by [a separate API](./history_and_modals.md).
+
+- _TODO: something something allowing trapping users on the page; need to figure out which interceptions we allow._
+
+- Provide an elegant layering onto or integration with the existing `window.history` API. That API is quite problematic, and we can't be tied down by a need to make every operation in the new API isomorphic to one in the old API.
+
+_TODO: figure out which of [Chris's goals and non-goals](https://docs.google.com/document/d/1eL7tOQXVw7hyV7tudS7WukT9qIsmZm0o-qGxq0wP2Ko/edit#) we haven't incorporated, but should._
+
+## Proposal
+
+### The current entry, and single-page navigations
+
+The entry point for the app history API is `window.appHistory`. Let's start with `appHistory.currentEntry`, which is an instance of the new `AppHistoryEntry` class. This class has the following readonly properties:
+
+- `key`: a user-agent-generated UUID identifying this history entry. In the past, applications have used the URL as such a key, but the URL is not guaranteed to be unique.
+
+- `url`: the URL of this history entry (as a string).
+
+- `state`: returns the application-specific state stored in the history entry (or `null` if there is none).
+
+- `sameDocument`: a boolean indicating whether this entry is for the current document, or whether navigating to it will require a full navigation (either from the network, or from the browser's back/forward cache). Note: for `appHistory.currentEntry`, this will always be `true`.
```

Comment:
Is it also interesting to know if a history entry is the start of a new document? You could have 10 entries, but only 3 documents. Is it useful to know which history entries belong to which document?

Reply:
I don't know; it might be! In general, I lack example code for how you would use many of these APIs, so I am having a hard time evaluating...